lib/courier/adapters/scheduler.ex

```diff
@@ -0,0 +1,153 @@
+      end
+
+      #lib/my_app/mailers/default.ex
+      defmodule MyApp.DefaultMailer do
+        use Courier, otp_app: :my_app
+      end
+
+      # config/dev.exs
+      config :my_app, MyApp.Mailer,
+        adapter: Courier.Adapters.Logger
+
+      # config/config.exs
+      config :my_app, MyApp.ScheduledMailer,
+        adapter: Courier.Adapters.Scheduler,
+        mailer: MyApp.DefaultMailer
+
+  To sent mail you just use `MyApp.ScheduledMailer.deliver/2`
+
+      Mail.build()
+      |> MyApp.ScheduledMail.deliver()
+
+  Courier will default to sending the email almost instantly. The assumed timestamp
+  when sending is `:calendar.universal_time/0`. You can tell Courier when the message should be sent
+  by passing the `at` option
+
+      tomorrow =
+        :calendar.universal_time()
+        |> :calendar.datetime_to_gregorian_seconds()
+        |> +(1 * 60 * 60 * 24) # 24 hours in seconds
+        |> :calendar.gregorian_seconds_to_datetime()
+
+      Mail.build()
+      |> MyApp.ScheduledMail.deliver(at: tomorrow)
+
+  The scheduler will delegate the message sending to the `mailer` declared in yor ScheduledMailer's config.
+
+  ## The Poller
+
+  The default polling interval is `0`. This is likely far too aggressive. To time interval for how frequently
+  the poller awakes to check for new messages to send simply set `interval` in the config:
+
+      # config/config.exs
+      config :my_app, MyApp.ScheduledMailer,
+        adapter: Courier.Adapters.Scheduler,
+        mailer: MyApp.DefaultMailer,
+        interval: 1_000 * 60 * 60  # awakes once an hour
+
+  The value for the interval is in milliseconds in accordance with the value that
+  [`:timer.sleep/1`](http://erlang.org/doc/man/timer.html#sleep-1)  expects.
+
+  ## Storage
+
+  The storage is where messages are kept until they are ready to be sent. The default
+  storage is `Courier.Storage.Simple` and is just an Agent, storing messages in-memory.
+  This may not be ideal for your use-case. You can override the storage in the config:
+
+      # config/config.exs
+      config :my_app, MyApp.ScheduledMailer,
+        adapter: Courier.Adapters.Scheduler,
+        mailer: MyApp.DefaultMailer,
+        storage: MyApp.OtherStorage
+
+  The custom storage will need respond to a certain API. Please see the documentation for `Courier.Storage`
+  for details or look at the source code for `Courier.Storage.Agent`.
+  """
+
+  @doc false
+  def deliver(%Mail.Message{} = message, opts) do
+    storage(opts).courier_put({message, timestamp(opts)})
+  end
+
+  @doc false
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  @doc false
+  def init(opts) do
+    {:ok, task_sup} = Task.Supervisor.start_link()
+    state = %{opts: opts, messages: %{}, task_sup: task_sup}
+    Process.send_after(self(), :poll, interval(opts))
+
+    {:ok, state}
+  end
+
+  @doc false
+  def handle_info(:poll, state) do
+    storage(state.opts).courier_all(past: true)
```

Comment:
We need to grab a handle to the `new_state` here after we reduce, otherwise add_message`'s are lost